colcon_cmake/event_handler/compile_commands.py

```diff
@@ -0,0 +1,121 @@
+                if workspace_level_json_path.exists():
+                    workspace_level_json_path.unlink()
+                return
+
+            # if the workspace-level json file is newer than all package-level
+            # json files it doesn't need to be regenerated
+            try:
+                workspace_level_mtime = os.path.getmtime(
+                    str(workspace_level_json_path))
+            except Exception:
+                pass
+            else:
+                for json_path in sorted(package_level_json_paths):
+                    try:
+                        mtime = os.path.getmtime(str(json_path))
+                    except Exception:
+                        continue
+                    if mtime > workspace_level_mtime:
+                        break
+                else:
+                    return
+
+            # collect all package-level json data
+            # for performance reasons avoid to load/dump json data
+            all_compile_commands = []
+            # keep deterministic order independent of aborted/selected packages
+            for json_path in sorted(package_level_json_paths):
+                compile_commands = self._get_compile_commands(json_path)
+                if compile_commands:
+                    all_compile_commands.append(compile_commands)
+
+            # generate workspace-level json file or remove if empty
+            if all_compile_commands:
+                with workspace_level_json_path.open('wb') as h:
+                    h.write(b'[\n')
+                    for i, compile_commands in enumerate(all_compile_commands):
+                        h.write(compile_commands)
+                        if i != len(all_compile_commands) - 1:
+                            h.write(b',\n')
+                        h.write(b'\n')
+                    h.write(b']\n')
+            elif workspace_level_json_path.exists():
+                workspace_level_json_path.unlink()
+
+    def _get_package_level_json_paths(self):
+        json_paths = set()
+        for package_name in self._package_names:
+            json_path = self._get_path(package_name)
+            if json_path.exists():
+                json_paths.add(json_path)
+        return json_paths
+
+    def _get_compile_commands(self, json_path):
+        content = json_path.read_bytes()
+        try:
+            open_index = content.index(b'[')
+            close_index = content.rindex(b'[')
+        except ValueError:
+            logger.warning(
+                "Data in '%s' is expected to be a list" % json_path.absolute())
+            return None
+        return content[open_index + 1:close_index - 1].strip()
+
+    def _get_path(self, package_name=None):
+        path = Path(self.context.args.build_base)
+        if package_name is not None:
+            path /= package_name
+        path /= CompileCommandsEventHandler.FILENAME
+        return path
+
+        self._file_handle = path.open(mode='w')
+        return True
```

Comment:
Isn't this unreachable? How did CI not catch it?

Reply:
Not sure what during CI should / would check for this. But yes, this is certainly unreachable. Removed in d365e4030cf5d79604b7730f7f199224fd1c1149.